core/src/main/java/org/jdbi/v3/core/extension/UseExtensionAnnotationHandlerCustomizer.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.extension;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.jdbi.v3.core.extension.annotation.ExtensionCustomizationOrder;
+import org.jdbi.v3.core.extension.annotation.UseExtensionCustomizer;
+import org.jdbi.v3.core.internal.JdbiClassUtils;
+import org.jdbi.v3.core.internal.exceptions.CheckedCallable;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Applies decorations to method handlers, according to any {@link UseExtensionCustomizer} decorating annotations
+ * present on the method. If multiple decorating annotations are present, the order of application can be controlled
+ * using the {@link ExtensionCustomizationOrder} annotation.
+ *
+ * @since 3.38.0
+ */
+final class UseExtensionAnnotationHandlerCustomizer implements ExtensionHandlerCustomizer {
+
+    static final ExtensionHandlerCustomizer HANDLER = new UseExtensionAnnotationHandlerCustomizer();
+
+    @Override
+    public ExtensionHandler customize(ExtensionHandler delegate, Class<?> extensionType, Method method) {
+        ExtensionHandler extensionHandler = delegate;
+
+        List<Class<? extends Annotation>> annotationTypes = Stream.of(method, extensionType)
+                .map(AnnotatedElement::getAnnotations)
+                .flatMap(Arrays::stream)
+                .map(Annotation::annotationType)
+                .filter(type -> type.isAnnotationPresent(UseExtensionCustomizer.class))
+                .collect(toList());
+
+        Stream.of(method, extensionType)
+                .map(e -> e.getAnnotation(ExtensionCustomizationOrder.class))
+                .filter(Objects::nonNull)
+                .findFirst()
+                .ifPresent(order -> annotationTypes.sort(createComparator(order).reversed()));
```

Comment:
this assumes `annotationTypes` is mutable, which `toList` does not guarantee. To guarantee a mutable list, you need to do e.g. `toCollection(new ArrayList<>())`